File: app/src/main/java/edu/mnstate/jz1652ve/lab27/MapsActivity.java
package edu.mnstate.jz1652ve.lab27;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentActivity;

import android.Manifest;
import android.content.pm.PackageManager;
import android.graphics.Camera;
import android.graphics.Color;
import android.os.Bundle;
import android.widget.Toast;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.UiSettings;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.CircleOptions;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;

public class MapsActivity extends FragmentActivity implements OnMapReadyCallback {
    private static final int LOCATION_REQUEST_CODE = 101;

    private GoogleMap mMap;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_maps);
        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager()
                .findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);

    }

    protected void requestPermission(String permission, int requestCode) {
        ActivityCompat.requestPermissions(this, new String[]{permission}, requestCode);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        switch (requestCode) {
            case LOCATION_REQUEST_CODE:
                if (grantResults.length == 0 || grantResults[0] != PackageManager.PERMISSION_GRANTED) {
                    Toast.makeText(this, "Can't show location - permission required", Toast.LENGTH_SHORT).show();
                }
                else {
                    SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager().findFragmentById(R.id.map);
                    mapFragment.getMapAsync(this);
                }
        }
    }

    /**
     * Manipulates the map once available.
     * This callback is triggered when the map is ready to be used.
     * This is where we can add markers or lines, add listeners or move the camera. In this case,
     * we just add a marker near Sydney, Australia.
     * If Google Play services is not installed on the device, the user will be prompted to install
     * it inside the SupportMapFragment. This method will only be triggered once the user has
     * installed Google Play services and returned to the app.
     */
    @Override
    public void onMapReady(GoogleMap googleMap) {
        final LatLng msumPos = new LatLng(46.8663522, -96.7620251);

        mMap = googleMap;

        UiSettings mapSettings = mMap.getUiSettings();

        mapSettings.setZoomControlsEnabled(true);
        mapSettings.setZoomGesturesEnabled(true);
        mapSettings.setScrollGesturesEnabled(true);
        mapSettings.setTiltGesturesEnabled(true);
        mapSettings.setRotateGesturesEnabled(true);

        Marker msum = mMap.addMarker(new MarkerOptions().position(msumPos).title("MSUM").snippet("Minnesota State University Moorhead"));

        mMap.addMarker(new MarkerOptions().position(new LatLng(0,0)).title("0x0").snippet("We be zeroed out"));
        mMap.addMarker(new MarkerOptions().position(new LatLng(45, 45)).title("Halfsies").snippet("Living on a prayer"));
        mMap.addMarker(new MarkerOptions().position(new LatLng(55.673611, 12.568333)).title("Tivoli").snippet("Have a blast"));

        CameraPosition camPos = CameraPosition.builder()
                .target(msumPos)
                .zoom(50)
                .bearing(70)
                .tilt(25)
                .build();

        mMap.animateCamera(CameraUpdateFactory.newCameraPosition(camPos));

        mMap.addCircle(new CircleOptions().center(msumPos).radius(500.0).strokeWidth(10f).strokeColor(Color.BLUE).fillColor(Color.argb(70, 10, 50, 150)));


        if(mMap != null) {
            int perm = ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION);
            if(perm == PackageManager.PERMISSION_GRANTED)
                mMap.setMyLocationEnabled(true);
            else
                requestPermission(Manifest.permission.ACCESS_FINE_LOCATION, LOCATION_REQUEST_CODE);

            mMap.setMapType(GoogleMap.MAP_TYPE_SATELLITE);
        }
    }
}

File: app/src/main/java/edu/mnstate/jz1652ve/finalproject/MapsActivity.kt
package edu.mnstate.jz1652ve.finalproject

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.location.Criteria
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.Menu
import android.view.View
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.location.LocationManagerCompat
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.Fragment
import com.google.android.gms.maps.*

import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.CircleOptions
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.material.navigation.NavigationView
import kotlinx.android.synthetic.main.activity_loc_picker.*
import java.util.jar.Manifest
import kotlin.math.sqrt

const val TAG = "MapsActivity"
private val LOC_PERM_ID = 101

class MapsActivity : AppCompatActivity(), OnMapReadyCallback {
    lateinit var locMan: LocationManager
    public var mMap: GoogleMap? = null
    public lateinit var nav: NavigationView
    public lateinit var drawer: DrawerLayout
    lateinit var onChange: () -> Unit
    val dao = ContactHelper(this)
    var desiredLoc: LatLng? = null

    private var curScreen: Int = R.id.navMap

    /**
     * Lets us commit a fragment async while also doing something when it's ready
     * without extra cruft
     */
    private fun <F : Fragment> setFragment(f: F, then: (F) -> Unit = {}) = supportFragmentManager
        .beginTransaction()
        .replace(R.id.mainFrag, f as Fragment)
        .runOnCommit {
            then(supportFragmentManager.findFragmentById(R.id.mainFrag) as F)
        }
        .commit()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        locMan = getSystemService(Context.LOCATION_SERVICE) as LocationManager

        setContentView(R.layout.activity_maps)

        setSupportActionBar(findViewById(R.id.toolbar))

        setFragment(MapFragment()) { frag: MapFragment ->
            this.onChange = { addMarkers() }
            frag.getMapAsync(this)
        }

        drawer = findViewById(R.id.root)
        nav = findViewById(R.id.menu)
        nav.setNavigationItemSelectedListener { item ->
            if (item.itemId != this.curScreen) {
                when (item.itemId) {
                    R.id.navMap -> {
                        setFragment(MapFragment()) {
                            it.getMapAsync(this)
                        }
                    }
                    R.id.navList -> {
                        setFragment(ContactListFragment.newInstance()) { frag ->
                            this.onChange = { frag.notifyChanged() }
                            frag.goToMap = { loc ->
                                curScreen = R.id.navMap
                                desiredLoc = loc
                                setFragment(MapFragment()) {
                                    it.getMapAsync(this)

                                }
                            }
                        }
                        mMap = null
                    }
                }
                this.curScreen = item.itemId
            }

            drawer.closeDrawer(GravityCompat.START)
            true
        }
    }

    override fun onMapReady(googleMap: GoogleMap) {
        mMap = googleMap
        mMap!!.mapType = GoogleMap.MAP_TYPE_HYBRID

        val perm = ContextCompat.checkSelfPermission(
            this,
            android.Manifest.permission.ACCESS_FINE_LOCATION
        )
        if (perm == PackageManager.PERMISSION_GRANTED)
            mMap!!.isMyLocationEnabled = true
        else
            askForgiveness(android.Manifest.permission.ACCESS_FINE_LOCATION, LOC_PERM_ID)

        if (desiredLoc != null) {
            mMap!!.moveCamera(CameraUpdateFactory.newLatLng(desiredLoc))
            desiredLoc = null
        }

        addMarkers()
    }


    @SuppressLint("MissingPermission")
    fun addMarkers() {
        if (mMap == null) return
        val map = mMap!!

        map.clear()

        val allContacts = dao.getAll()
        if (map.isMyLocationEnabled)
            locMan.requestSingleUpdate(LocationManager.NETWORK_PROVIDER, object : LocationListener {
                override fun onLocationChanged(location: Location?) {
                    val otherLoc = Location(location)
                    var furthest = 0.0f
                    for (c in allContacts) {
                        otherLoc.latitude = c.lat
                        otherLoc.longitude = c.lng
                        val d = location!!.distanceTo(otherLoc)
                        if (d > furthest) furthest = d
                    }

                    map.addCircle(
                        CircleOptions()
                            .center(LatLng(location!!.latitude, location!!.longitude))
                            .radius(furthest.toDouble())
                            .strokeColor(Color.BLUE)
                            .fillColor(Color.argb(20, 0, 0, 255))
                    )
                }

                override fun onStatusChanged(provider: String?, status: Int, extras: Bundle?) {}
                override fun onProviderEnabled(provider: String?) {}
                override fun onProviderDisabled(provider: String?) {}
            }, null)





        for (c in allContacts) {
            val pos = LatLng(c.lat, c.lng)
            map.addMarker(
                MarkerOptions()
                    .position(pos)
                    .title(c.firstName + " " + c.lastName)
                    .snippet(c.phone)
            )
        }
    }

    fun onAddBtn(view: View) {
        var goAddItBoy = Intent(this, AddContactActivity::class.java)
        this.startActivityForResult(goAddItBoy, 0)
    }

    private fun askForgiveness(perm: String, id: Int) {
        ActivityCompat.requestPermissions(this, arrayOf(perm), id)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        when (requestCode) {
            LOC_PERM_ID -> {
                if (grantResults.isEmpty() || grantResults[0] != PackageManager.PERMISSION_GRANTED)
                    Toast.makeText(
                        this,
                        getString(R.string.sphereDisabled),
                        Toast.LENGTH_SHORT
                    ).show()
                else if (mMap != null) { // i.e we're in the map fragment at the moment
                    val mapFragment =
                        supportFragmentManager.findFragmentById(R.id.mainFrag) as MapFragment
                    mapFragment.getMapAsync(this)
                }
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        this.onChange()
    }
}